staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go

```diff
@@ -315,13 +317,18 @@ func (le *LeaderElector) tryAcquireOrRenew() bool {
 		return true
 	}
 
+	if reflect.DeepEqual(le.observedRecord, rl.LeaderElectionRecord{}) {
+		firstStart = true
+	}
+
 	// 2. Record obtained, check the Identity & Time
 	if !reflect.DeepEqual(le.observedRecord, *oldLeaderElectionRecord) {
 		le.observedRecord = *oldLeaderElectionRecord
 		le.observedTime = le.clock.Now()
 	}
 	if len(oldLeaderElectionRecord.HolderIdentity) > 0 &&
 		le.observedTime.Add(le.config.LeaseDuration).After(now.Time) &&
+		!(firstStart && oldLeaderElectionRecord.RenewTime.Add(le.config.LeaseDuration).Before(now.Time)) &&
```

Comment:
This time stamp is not meaningful if it was collected on another machine. The implementation of this client only acts on locally collected time stamps and cannot rely on the accuracy of time stamp in the record for correctness. The client needs to wait the full lease duration without observing a change to the record before it can attempt to take over. Start isn’t different. This is documented here:

https://github.com/kubernetes/kubernetes/blob/f5a1ceb1fcb3572d37865da8257f66659f8004a9/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L22-L24

10 minute lease duration sounds too long. It’s defaulted to 15 seconds in core components:

https://github.com/kubernetes/kubernetes/blob/c48ed1c5aca6e06b6a3d70d5f189216df3150379/cmd/cloud-controller-manager/app/options/options_test.go#L54

Reply:
@mikedanese So it is designed by purpose not a bug, actually?

Reply:
> This time stamp is not meaningful if it was collected on another machine. The implementation of this client only acts on locally collected time stamps and cannot rely on the accuracy of time stamp in the record for correctness. The client needs to wait the full lease duration without observing a change to the record before it can attempt to take over. Start isn’t different. 

This absolutely needs to be added to the leader election package documentation, IMO. It is more cleaer in describing a actual scenario.

Reply:
Yes the intended behavior:

https://github.com/kubernetes/kubernetes/blob/3d12466c0221549ffd0651668e078b53482bcad2/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L105-L108

And "A client observes timestamps captured locally to infer the state of the leader election" are both documented but an example would definitely make this clearer. We should also document reasonable values for LeaseDruation, RenewDeadline and RetryPeriod.

An ack is any change of the record observed by the client. The RenewTime could change to the previous day, and it would still reset the LeaseDuration countdown for non-leader clients waiting to become leader.

For a starting process, the countdown should start at the time that the initial record is read.

Reply:
@mikedanese In a distributed system, time coherence in different nodes are hard.

Reply:
@mikedanese I have updated the PR title and content to add more doc about this behavior. PTAL.